Add AMDGPU-PRO driver

corngood opened the pull request:
###### Motivation for this change

This adds the latest [proprietary AMD gpu drivers](http://support.amd.com/en-us/kb-articles/Pages/AMD-Radeon-GPU-PRO-Linux-Beta-Driver%e2%80%93Release-Notes.aspx), which include a vulkan driver.  It's loosely based on the ati proprietary package.  The patches for the kernel module come from the gentoo/arch packages, and allow it to be built for linux 4.4-4.7.  It looks like 4.8 will require more patches.

The driver can be activated by setting:

```
services.xserver.videoDrivers = [ "amdgpu-pro" ];
```
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- [x] Test 64-bit OpenGL
- [x] Test 64-bit Vulkan
- [ ] Test 64-bit OpenCL
- [x] Test 64-bit vdpau
- [x] 32-bit build
- [x] Test 32-bit OpenGL
- [x] Test 32-bit Vulkan
- [ ] Test 32-bit OpenCL
- [x] Test 32-bit vdpau

---

corngood committed:
amdgpu-pro: Init at 16.30.3-315407

corngood commented:
Updated with 32-bit build, vulkan and opengl tested so far.

corngood commented:
Tested vdpau 32/64.  It works, but seems a bit unstable on my R9 290.

Anyone have a suggestion for a good opencl test program?

domenkozar commented:
@corngood best to also specify how to test the drivers :)

vcunat commented:
Hmm, the "-pro" version should use a different kernel module than the fully open-source amdgpu solution? I thought AMD originally announced the kernel stuff would be shared?

vcunat commented:
BTW, I certainly don't consider OpenCL a blocker for merging; testing the rest seems good enough.

corngood commented:
@domenkozar Good point.  Config snippet added to the top comment.

corngood commented:
@vcunat Well, the kernel module seems to be a similar to what's in linux, with changes needed to build with dkms for linux-4.4, and some new features and device support that haven't been merged yet.

The proprietary user-space actually sort-of works with the amdgpu driver in linux-4.7, but there are some nasty surprises: notably, fences don't work in Vulkan.

It would be nice if the kernel module was called amdgpu-pro.ko, but is it worth the trouble and extra patches?

corngood commented:
I think I need to rebase to fix these build errors.  Also, I suppose I should add myself as maintainer.

vcunat commented on pkgs/os-specific/linux/amdgpu-pro/default.nix:
```diff
@@ -0,0 +1,146 @@
+
+  kernelDir = if libsOnly then null else kernel.dev;
+
+  driverLink = "/run/opengl-driver" + optionalString stdenv.isi686 "-32";
```
It will be cleaner to have `inherit (mesa_noglu) driverLink`.

vcunat commented on pkgs/os-specific/linux/amdgpu-pro/default.nix:
```diff
@@ -0,0 +1,146 @@
+    homepage =  http://support.amd.com/en-us/kb-articles/Pages/AMDGPU-PRO-Beta-Driver-for-Vulkan-Release-Notes.aspx ;
+    license = licenses.unfree;
+    platforms = platforms.linux;
+    hydraPlatforms = [];
```
Better drop the `hydraPlatforms` line, unless there's some other reason than `license`. Hydra does license filtering by itself.

vcunat commented:
>  I think I need to rebase to fix these build errors.

Travis for nixpkgs seems completely broken ATM, if you mean that.

> Also, I suppose I should add myself as maintainer.

That would be good, especially if you would generally support user's problems with the driver, so they would know whom to ping. Hydra won't build unfree stuff, so you can't get build failure notifications...

corngood committed:
Add myself as maintainer

corngood commented:
@vcunat I updated it according to your comments.

I also noticed that the binary replacement (using perl) was fragile, and for the 32-bit libs it was writing past the end of the string.

I switched it to just replace /usr/lib/? with /run/lib/? (with an assert that the lengths match) and added an activation script to create those symlinks.  I'm not really happy with this, but I'm not sure of a less hacky way to do it.

vcunat commented:
Hmm, such replacements in binary files are a little risky IMO. Do you know how this `/usr/lib` is used within? If it was just RPATH etc., `patchelf` could solve it all in a much better way, without needing that symlink, too.

corngood commented:
Yeah, it's a last resort.  It's not RPATH..  As far as I can tell it's stuff like DRI_DRIVER_INSTALL_DIR, which comes from autoconf and is compiled to a string literal.  It's possible that some of it can be overridden by environment variables, but that's not very nix-friendly either.

Now that the replaced paths are guaranteed to be the same length, it should be pretty safe (/run/lib vs /usr/lib).  Is this an ok way of using /run/?  It seems better than trying to create a /usr/lib, if that's even possible.

The problematic replacement was:

```
/usr/lib/i386-linux-gnu => 
/run/opengl-driver-32/lib
```

Even replacing with /run/ paths is kind of hacky.   It should really be using /nix/store/, but obviously those are far too long.

corngood commented:
I think this is in good enough shape to get out there for more testing.  Let me know if you have any more comments.

vcunat commented:
Question not relevant to merging this: have you/someone heard of glvnd support in amdgpu* or fglrx? I've found no relevant references, so I can only suppose they don't plan to support it soon.

vcunat commented:
When trying to build for 32-bit kernel, it fails with:
```
installing
rm: cannot remove '/nix/store/8yjdr38353v3yrbp0ig08b4sjd3spa4i-amdgpu-pro-16.30-4.4.23/lib/xorg': No such file or directory
```

but I suppose that's such a rare combination that we could just mark it `meta.broken` or something.

corngood commented:
I haven't seen any indication of glvnd support in the drivers, but AMD did promise it a while back.

As for the 32-bit thing, I guess it doesn't ship with 32-bit xorg drivers, which makes the kernel module pretty useless, even if it does compile.  It could skip them, but it would just fail on xorg startup, so asserting `libsOnly` for i686 is probably best.

corngood commented:
@vcunat Sorry for the delay, I've been traveling.

Instead of disabling anything I just added guards for the missing xorg drivers.  At least you should get a working drm driver on i686 which will work with the kms xorg driver.  Also if amd decides to ship a 32-bit xorg driver in the future it should 'just work'.

vcunat commented:
I was out of time for about a week, until a while ago, so your timing is perfect :-)

Pushed to 16.09 as well.

corngood commented:
Thanks.  Is there anything I can do to get notifications about issues in this package, or is it up to reporters/triage to manually tag maintainers?

vcunat commented:
Nothing else I know of. People should mention maintainers. The bot would probably automatically mention you on PRs touching these files.
